tremor-script/src/ast/deploy/raw.rs

```diff
@@ -0,0 +1,785 @@
+        let run_consts = run_consts.run();
+        let env = Env {
+            context: &ctx,
+            consts: run_consts,
+            aggrs: &$helper.aggregates.clone(),
+            meta: &$helper.meta.clone(),
+            recursion_limit: 1024, // I'm feeling lucky in a power of two kind of way - shoot me
+        };
+        let got = $expr.run(eo, &env, &mut event, &mut state, &mut meta, &mut local);
+        match got {
+            Ok(Cont::Emit(value, _port)) => value,
+            Ok(Cont::Cont(value)) => value.into_owned(),
+            // FIXME this deserves a proper error
+            _otherwise => {
+                return error_generic(
+                    $expr,
+                    $expr,
+                    &"Illegal `with` parameter value specified".to_string(),
+                    &$helper.meta,
+                )
+            }
+        }
+    }};
+}
+
+/// Convenience to evaluate an expression at compile time so that
+/// evaluated results can be used during compilation
+#[macro_export]
+macro_rules! stateless_run_str {
+    ($helper:ident, $expr:expr) => {{
+        let eo = ExecOpts {
+            aggr: AggrType::Emit,
+            result_needed: true,
+        };
+        let ctx = EventContext::new(nanotime(), None);
+        let event = literal!({}).into_static();
+        let state = literal!({}).into_static();
+        let meta = literal!({}).into_static();
+
+        let local = LocalStack::with_size(0);
+
+        let run_consts = $helper.consts.clone();
+        let run_consts = run_consts.run();
+        let env = Env {
+            context: &ctx,
+            consts: run_consts,
+            aggrs: &$helper.aggregates.clone(),
+            meta: &$helper.meta.clone(),
+            recursion_limit: crate::recursion_limit(),
+        };
+        $expr.run(eo, &env, &event, &state, &meta, &local)
+    }};
+}
+
+pub(crate) fn up_params<'script, 'registry>(
+    params: &[(IdentRaw<'script>, ExprRaw<'script>)],
+    helper: &mut Helper<'script, 'registry>,
+) -> Result<HashMap<String, Value<'script>>> {
+    let mut mapped = HashMap::new();
+    for (name, value) in params {
+        let name = name.clone().up(helper)?.to_string();
+        let expr = value.clone().up(helper)?;
+        let value = stateless_run_expr!(helper, &expr).clone_static();
+        mapped.insert(name, value);
+    }
+
+    Ok(mapped)
+}
+
+pub(crate) fn up_maybe_params<'script, 'registry>(
+    params: &Option<DeployWithExprsRaw<'script>>,
+    helper: &mut Helper<'script, 'registry>,
+) -> Result<Option<HashMap<String, Value<'script>>>> {
+    params
+        .as_ref()
+        .map(|params| up_params(params, helper))
+        .transpose()
+}
+
+#[derive(Debug, PartialEq, Serialize)]
+#[allow(clippy::module_name_repetitions)]
+pub struct DeployRaw<'script> {
+    pub(crate) config: WithExprsRaw<'script>,
+    pub(crate) stmts: DeployStmtsRaw<'script>,
+}
+impl<'script> DeployRaw<'script> {
```

Comment:
This looks like it could be handled by `Uppable` making it reusable in all the places we use that trait

Reply:
Now why didn't i think of this!